Setup remote Github repo if username available

rffrancon opened the pull request:
Currently, users must setup a Github repository using the Github website before trying out crev. This PR should include changes which would allow a user to create a remote Github repo from crev directly using `cargo crev new id`.

The Github username can be given directly or extracted from a github repo URL.

This isn't perfect but should make it easier for people to try out crev.

dpc commented on crev-lib/src/local.rs:
```diff
@@ -249,38 +249,109 @@ impl Local {
+            repo.remote_set_url("origin", &git_https_url)?;
+        }
+
+        let args: Vec<OsString> = vec![
```
Doesn't this require Auth of some kind?

dpc replied:
I wasn't even aware of this API. Good to know!

dpc commented:
I'm not sold on this. I'll comment more when I can, but it adds complexity, possibly breaks some use cases and the improvement is not that big.

Dylan-DPC commented:
>  which would allow a user to create a remote Github repo from crev directly 
this could be good intiially for trying out crev, but overall i don't see why would someone need to run crev if they don't own a repo.

dpc commented:
So what happens when the user already do have a repo (you can use the same repo on multiple computer, potentially for different IDs too)? Or if they don't remember the password, or like me use only yubikeys for ssh access? Or if they want a different name of the repo?

There is too many cases here and "give us an url to a repo that already exists" is a simple unification that both removes the complexity from `crev` itself, documentation, and UX.

rffrancon commented:
I believe that all of the cases that you've mentioned are properly handled. This new routine only takes effect if cloning from a given git URL fails. See: https://github.com/dpc/crev/pull/62/commits/c293395c36262b466fabbb3b9dbd88aea0a4c791#diff-3ea02b09d8e55e73368efe3739af5964R342

Therefore, given a git URL or a username, an attempt is firstly made to clone from the expected repo.

dpc commented:
I have missed that. I looked through it again, but left the comments in the wrong place :https://github.com/dpc/crev/commit/c293395c36262b466fabbb3b9dbd88aea0a4c791#diff-3ea02b09d8e55e73368efe3739af5964R342

dpc commented:
So if this is just a fallback, I guess I'm OK with this. Notes:

* if we fail to clone, we should initialize empty dir, set url like before
* then we should parse the provided url (using a generalized function based on `https_to_git_url`) , and if we detect that domain was `github.com` and rest of the url has `<username>/<reponame>` then ask user explicitly if they want to attempt to create the repo automatically. I just want to avoid `cargo crev` suddently asking for user passwords.
* Avoiding calling `curl` and `git` binaries would be a plus.

rffrancon commented:
@dpc Thank you kindly for the review and the feedback. I really appreciate it.
I will address your comments and then rebase and force push onto this branch.

You've raised many points which I'll need to think about carefully. But just quickly on the issue of `curl`. I agree with what you've said. Would you accept including this crate: https://github.com/alexcrichton/curl-rust ? I'm not certain if it will work in this instance, but it probably will.

rffrancon commented:
@Dylan-DPC Thank you for your comments. I've addressed them, left a question on one.

@dpc I believe that I've addressed your comments. However, I did not implement the new git command in `git2-rs`, I think the commit is getting too large. Please let me know what you think of the latest changes.

dpc commented on crev-lib/src/lib.rs:
```diff
@@ -196,35 +196,49 @@ where
+
+    let (git_https_url, github_username) = parse_url_and_username(&git_url_or_github_username);
+    eprintln!("Repository URL: {}\n", git_https_url);
+    eprintln!("It is recomended that this repository exists and is initialized upfront (can be empty).");
```
Why no loop? What if user made a typo? :)

Please don't change this part. Just yesterday I went through  https://github.com/dpc/crev/issues/75 to address it. If you think it should be changed/improved, let's discuss it slowly in a separate issue and PR.

dpc replied:
I am actually going to scrap the whole thing and change it. There is going to be no interactive prompts or anything, except for asking for a pasword. So please revert this part.

dpc commented on crev-lib/src/local.rs:
```diff
@@ -60,12 +73,70 @@ fn https_to_git_url(http_url: &str) -> Option<String> {
+}
+
+fn https_to_git_url(http_url: &str) -> Option<String> {
+    match parse_git_url_https(http_url) {
```
Looks like you want a `.map(...)`

dpc replied:
X

dpc commented on crev-lib/src/local.rs:
```diff
@@ -88,6 +159,74 @@ fn https_to_git_url_test() {
 }
 
+fn parse_git_repository_name(git_https_url: &str) -> Option<String> {
+    let parts = git_https_url.split('/');
```
This could just call `parse_git_url_https` instead, no? No need to have that logic multiple times.

dpc replied:
X

dpc commented on crev-lib/src/local.rs:
```diff
@@ -88,6 +159,74 @@ fn https_to_git_url_test() {
+    }
+
+    match parts.last() {
+        Some(last_part) => {
```
`.map` :)

dpc replied:
X

dpc commented on crev-lib/src/local.rs:
```diff
@@ -243,37 +382,108 @@ impl Local {
 
-        let proof_dir =
-            self.get_proofs_dir_path_for_url(&Url::new_git(git_https_url.to_owned()))?;
+        let readme_file_name = Path::new("README.md");
```
I've just added a function creating `README_USING_THIS_REPO.md`. Use it instead. `README.md` is supposed to  be personalized, for the user.

dpc commented:
You know what? I'm going to do it all differently. This whole interactive stuff is just annoying, and confusing. 

If you don't mind I'm going to manually take your code and turn it into `--create` switch to `cargo crev new id`. If you want can you address the issues that I'm going to mark with X.

dpc commented:
There reason I'm going to do it differnetly, is that it's supposed to be a command line tool, and not an interactive console RPG game.

rffrancon commented:
Thank you for looking over my changes once more. I've addressed your comments. I agree with you wholeheartedly on moving away from the interactive side! :) The interface you've suggested looks good. This should also cleanup stdout, eventually help messages could be moved to the help command.

Good stuff!

dpc commented:
I've got it to work, and even pushed to master. But ultimately decided this too clunky, and too much problem for what it's worth. People shouldn't really be using http auth and entering password to their github accounts like this. It's all just bad idea. So sorry, but I don't think this is worth implementing, after all.

rffrancon commented:
@dpc I understand. Thank you for your reviews regardless.
